src/CMakeLists.txt

```diff
@@ -1,60 +1,85 @@
+# find required packages for base and app
+find_package(LibtorrentRasterbar QUIET ${requiredLibtorrentVersion} COMPONENTS torrent-rasterbar)
```

Comment:
Sorry for repeating; this chunk has to be packed in a FindLibtorrentRasterbar.cmake file and put in the Modules dir. Distributions should be able to their own versions of Find packages.

Reply:
@zeule in that case, will it still be used preferentially over any other `FindLibtorrentRasterbar.cmake` in the system unless the user overrides that? Or is there something more that I must do to obtain that behaviour?

Reply:
No, that the point: the system one has to be preferred.

Reply:
@zeule

> the system one has to be preferred.

I don't see why. This code does not do anything special to find the package, it just finds the config file and lets that do the magic. The system is free to provide/give preference to whatever `LibtorrentRasterbarConfig.cmake` it sees fit. I think this is the sanest way: do as little as possible, trust the provided config files (because this is most likely the best from the system).

So IMO our way of _finding_ should be the default. The system's way of finding should be the fallback (automatically), but the user should also be able to override this order and prefer the system's way of finding regradless, via a variable or possibly a feature.

The fallback is useful for cases where, for example, libtorrent is not used by CMake.

Reply:
Libtorrent might be built with b2, autotools or cmake. The config file will be installed only by the latter one, but that's a side note. 

For a distribution it is much easier to patch the Find module file which comes with the libtorrent itself then duplicate required changes in each client package. If you want to make their life easier (in other words: if you want them to prefer CMake build for qBt) you put the _whole_  find chunk into a module and leave only a single `find_package()` call in the build script.

Reply:
>Distributions should be able to provide their own version of Find packages.

Seems reasonable. Distribution can provide more reliable way to find its packages.

Reply:
I don't understand this whole thing. The system FindLibtorrentRasterbar.cmake is already being used, this code merely says "if FindLibtorrentRasterbar.cmake isn't found, try to use pkg-config instead".

Why is this objectionable? When does this prevent the distribution from providing their own FindLibtorrentRasterbar.cmake or patching the system one?

Reply:
>FindLibtorrentRasterbar.cmake 

The point is that all "find libtorrent" logic should be encapsulated in FindLibtorrentRasterbar.cmake (either system or project, user etc.) so there is single find_package call should be in project file.

Reply:
This is some pretty abstract, confusing logic. If I understand you correctly, then these lines here:

```
find_package(Boost ${requiredBoostVersion} REQUIRED COMPONENTS system)
find_package(OpenSSL ${requiredOpenSSLVersion} REQUIRED)
find_package(ZLIB ${requiredZlibVersion} REQUIRED)
```

Should be moved to modules called e.g.
```
$ cat FindZLIB.cmake
find_package(ZLIB ${requiredZlibVersion} REQUIRED)
```

and invoked as find_package(ZLIB) to resolve ./cmake/Modules/FindZLIB.cmake which in turn resolves /usr/share/cmake-3.17/Modules/FindZLIB.cmake. But in that case, find_package(ZLIB) is also a module lookup functionality, so it too should be moved to a module in cmake/modules/. How many times do we recursively factor this out? What's the objective in doing so repeatedly? What's the objective in your proposal to do so even once?

If the objective here is to make distros able to override the functionality, then since a FindLibtorrentRasterbar.cmake is *always* used, they just provide one and have done with. You can factor this out *zero* times as in the current PR and it still works as expected.

If you're introducing additional motivations that @zeule doesn't have, please say so as I'm a bit confused here.

Reply:
> The point is that all "find libtorrent" logic should be encapsulated

This.

Reply:
So just to be clear:

> Distributions should be able to provide their own version of Find packages.

> No, that the point: the system one has to be preferred.

> For a distribution it is much easier to patch the Find module file which comes with the libtorrent itself then duplicate required changes in each client package. If you want to make their life easier (in other words: if you want them to prefer CMake build for qBt) you put the whole find chunk into a module and leave only a single find_package() call in the build script.

is all actually completely wrong and you're retracting those statements? i.e. This is about the neatness of factoring internal logic into separate, composable files?

Meanwhile anyone who is interested in changing how qbittorrent is detecting libtorrent-rasterbar will NOT be modifying ./cmake/modules/*.cmake, they will be modifying /usr/share/cmake-3.17/Modules/*.cmake? or somewhere else on the cmake modules path which is external to qbittorrent?

Reply:
> Distributions should be able to provide their own version of Find packages.

True.

> …the system one has to be preferred.

True.

> For a distribution it is much easier to patch the Find module file which comes with the libtorrent…

True.

> all the "find libtorrent" logic should be encapsulated

True.

> Meanwhile anyone who is interested in changing how qbittorrent is detecting libtorrent-rasterbar … will be modifying … cmake modules path which is external to qbittorrent?

True.

Reply:
@zeule @eli-schwartz 

> For a distribution it is much easier to patch the Find module file which comes with the libtorrent itself then duplicate required changes in each client package.

Am I missing something? You don't need to duplicate changes. I see 2 cases:

- the distro can provide a single `FindLibtorrentRasterbar.cmake` for all possible client packages. It does not matter if it is patched or not or shoots fireworks. This code will find it and use it.
- the distro has to provide a certain way of finding libtorrent for package A, and a different, possibly mutually exclusive way, for finding it for package B, independently of the default way. In that case, wouldn't they have to put each of the different `FindLibtorrentRasterbar.cmake` into each package's CMake modules dir, so that it gets picked up in preference to the system's?
This code will find it in the project's CMake modules and use it preferentially.

In either case if finding fails at first, `pkgconfig` is tried next.

Reply:
@FranciscoPombal if you do not understand why encapsulation is important I see a little sense to interfere. To give you a hint, that bullshit which you wrote with faked packages will scary off sane persons and it does not even work properly.

Reply:
>In either case if finding fails at first, pkgconfig is tried next.

Why do you separate it? `pkgconfig` is yet another "find package" way. "finding fails" means (should mean!) CMake cannot find package in all defined ways (including `pkgconfig` way if it is supposed to be used).

Reply:
@zeule

>To give you a hint, that bullshit which you wrote with faked packages will scary off sane persons and it does not even work properly.

First off, maybe other people you know write bullshit, I don't. I'm human though, so sometimes I make mistakes or do things in a sub-optimal way. When that happens, I expect people to provide constructive criticism, so that I may learn with the mistake and do better next time. So far I have solved every review you posted, and taken your suggestions into account. Sometimes I was doing things the wrong way, you suggested a better one, I recognized it, and did it that way (the `ALIAS` thing, for example). That's how it is supposed to work.

So how about actually explaining calmly what is wrong with that part? In my testing it seemed to work well. My instinct does tell me it feels rather hacky, but it is the simplest way I found to get `libtorrent` to show up in the feature summary when found by pkgconfig. Otherwise, the only indication that pkgconfig gives that it found `libtorrent` is a single output line in the beginning (near the "Found working compiler" noise that "disappears" into scrollback). So please, explain where/when exactly it does not work and/or how it can be done better, instead of insulting, because I think nobody wants to go down that road.

> if you do not understand why encapsulation is important I see a little sense to interfere. 

Again, unnecessary insult.
I do understand encapsulation and want to move towards it. If this is still about our previous discussion, I have already told you that truly encapsulating the logic in qBittorrent's CMake scripts will happen in a later PR. There are enough changes as it is for now, it will already be a pain in the ass to organize this neatly into separate commits just to have a clean and clear structure of commits. That being said, it's not like this PR doesn't move towards more encapsulated logic at all, anyway, it does. It's just that the big push for it has to come later, after a solid foundation is laid out.

If you're convinced that I don't understand the principle of encapsulation, then I'm convinced you don't understand the principle of least surprise. Originally, you defended the "find package X -> enable feature Y as a side-effect" practice by claiming that's how KDE (or some other high-profile project using Qt + CMake) does it (for example, ẁith the `Qt5Widgets` package and a GUI feature).
It's good to take inspiration from "authoritative" sources, but one must always consider some "done things" may not be best practices even in these projects. All projects have technical debt, and sometimes they survive and thrive in spite of certain design decisions, not because of them. I believe this is the case for "features by package found side-effect".
qBittorrent's CML is moving on from that paradigm, and I believe everyone is convinced it is technically the better option. Even in libtorrent, recently, the topic resurfaced, and is to be handled at a later date.
It is obvious you have great experience with CMake, but you are not impervious to making mistakes.
In the example `FindLibtorrentRasterbar.cmake` file you wrote (in the libtorrent repo), there is this:

```cmake
		# Without pkg-config, we can't possibly figure out the correct build flags.
		# libtorrent is very picky about those. Let's take a set of defaults and
		# hope that they apply. If not, you the user are on your own.
		set(LibtorrentRasterbar_DEFINITIONS
			-DTORRENT_USE_OPENSSL
			-DTORRENT_DISABLE_GEO_IP
			-DBOOST_ASIO_ENABLE_CANCELIO
			-D_FILE_OFFSET_BITS=64)
```

More surprises. The user doesn't even get warned that this happened. They would have to go look up this comment. IMO the more sane option would be to fail and inform the user that they should manually pass something like `-DLibtorrentDefs=...` and just check for the existence of that variable. Or just tell them to fix their pkgconfig config/CMake config files and try again.

@glassez 

Why do you separate it? `pkgconfig` is yet another "find package" way. "finding fails" means (should mean!) CMake cannot find package in all defined ways (including `pkgconfig` way if it is supposed to be used).

The goal is to only use the CMake config files like `LibtorrentRasterbar.config` first, and then pkgconfig if necessary. If the distro has a special need not covered by this they are free to write their own `FindLibtorrentRasterbar.cmake`, put it in the cmake/Modules folder, and that will be used instead. I don't think _we_ should concern ourselves with that.

And besides, what exactly is the case where a distro packages libtorrent, but in such a way that it does not include the CMake config file _and_ doesn't work with pkgbuild, anyway?

Reply:
>The goal is to only use the CMake config files like LibtorrentRasterbar.config first, and then pkgconfig if necessary. 

I'm sorry, I'm tired of standing in one place... I really don't understand why you don't want to encapsulate all this logic in single module as it's expected?

Reply:
@glassez I would also rather not be stuck in this discussion, but here we are.

> I really don't understand why you don't want to encapsulate all this logic in single module as it's expected?

I don't see what logic there is to encapsulate. The current code just trusts whatever `LibtorrentRasberbar.config` or `FindLibtorrentRasterbar.cmake` it finds in the system. Those are the files that do the actual heavy lifting in the sense of encapsulating the logic of finding out definitions, compile options, etc, and other metadata of the library that's needed to successfully link to it.

Presumably, a distribution will include `LibtorrentRasberbarConfig,cmake` in their libtorrent package, so that will be used. If they have special needs, they can provide a specific `FindLibtorrentRasterbar.cmake` with the qBittorrent package, and that will be used instead. If none of the previous two cases applies, pkgconfig is used to discover the required metadata.

The current code is not a find script nor pretends to be one; it just says "try to use an actual find script/config, and if that does not work, try pkgbuild". At most, this code could be factored out to a separate macro or function. Calling it and treating it like find script logic is misleading IMO.

Furthermore, here is what I think, related to my second bullet point here: https://github.com/qbittorrent/qBittorrent/pull/12746#discussion_r426332927, which has not been addressed thus far: Imagine if every client package of `libtorrent` provided their separate `FindLibtorrentRasterbar.cmake`. Over time, they would diverge, some would be somewhat unmaintained, and it would be hell for the distro maintainer to try to figure out how exactly to package libtorrent in such a way that no client package breaks. Or they'd have to patch all the find scripts, carefully. On the other hand, if the onus of providing a good and flexible `LibtorrentRasberbarConfig,cmake` or pkgconfig metadata falls on the library itself, the logic on the client packages can be much simpler (boils down to "trust the cmake config files/pkgconfig metadata you find in the system"), and the maintenance burden is lightened. The maintainer just has to patch the client packages that don't work with what the library provides by default.
Fortunately, I find that as far as libtorrent is concerned, the generated CMake config files/pkgconfig metadata are quite good.

Reply:
@glassez,

pkg_check_modules() is an **alternative** to find_package(), they're two completely different ways to define a cmake "imported target". Either one under the hood simply calls add_library(... IMPORTED) and sets interface properties on it.

While it could, I suppose, be `include()`d from cmake/Modules/, it doesn't seem unambiguously obvious to do so, and this entire discussion is explicitly about distributing it as a "Find" package anyway.

Reply:
@FranciscoPombal, everyone, I believe, has a right to produce bullshit and name their own or somebody else one their liking.

> So how about actually explaining calmly what is wrong with that part? 

It fails if PkgConfig finds no libtorrent package, it fails if libtorrent version is too new and unsupported. What's wrong with it is that it creates problems because you don't want to follow the standard way for locating packages.

I maybe don't have enough experience with large software built with CMake, but what I learned from more than 10 years of running Gentoo on my laptops is that a FOSS project build system is not the place to exercise abilities to create and invent. Builds happen in unpredictable environments and fail more often that developers think. Those who install software have no time or crave to study another non-standard build script to compile that piece of software when something does not work. Because this is not the type of think explained in a wiki or readme (unlike build options). If projects follow similar patterns life of those who install software can be much easier. Find modules for CMake is one of the strongest patterns of that kind.


> ...it is the simplest way I found to get `libtorrent` to show up in the feature summary when found by pkgconfig.

Find module + `find_package_handle_standard_args()` is the simplest way.

> I do understand encapsulation and want to move towards it. 

Then do it, please.

> I have already told you that truly encapsulating the logic in qBittorrent's CMake scripts will happen in a later PR. There are enough changes as it is for now, 

... ruining encapsulation :)

> If you're convinced that I don't understand the principle of encapsulation, then I'm convinced you don't understand the principle of least surprise.

Totally controlled by ones preexisting experience. I built KDE for god knows how many times, that had its impact, perhaps. 
 
> ```cmake
> 		# Without pkg-config, we can't possibly figure out the correct build flags.
> 		# libtorrent is very picky about those. Let's take a set of defaults and
> 		# hope that they apply. If not, you the user are on your own.
> 		set(LibtorrentRasterbar_DEFINITIONS
> 			-DTORRENT_USE_OPENSSL
> 			-DTORRENT_DISABLE_GEO_IP
> 			-DBOOST_ASIO_ENABLE_CANCELIO
> 			-D_FILE_OFFSET_BITS=64)
> ```
> 
> More surprises. 

Well, I use Windows only to launch Steam. Have no idea what are the best practices there.

Reply:
@zeule You're doing a truly terrible job of communicating your intentions.

I told you I was interpreting your statements as mutually contradictory, asked you to clarify your meaning as to what "the point" is, and you simply replied with some "True. True. True." gimmick which I can only assume is supposed to be terribly witty.

And apparently you also cannot stand to explain yourself to @FranciscoPombal either, so once again instead of taking the opportunity of having a discussion to, well, "discuss" things, you just smear insults across the screen. I'm left wondering why you bother to be here.

...

I still do not understand your first argument in favor of splitting out the "try to find an external provider of libtorrent-rasterbar" logic, because you say it is in aid of linux distributions (or other consumers, one must suppose) replacing the logic with their own logic. This is, to my understanding, nonsense, because anything used to replace said logic would trigger the find_package() invocation. On the other hand, it is nonsense anyway, because I AM A LINUX DISTRIBUTION (I package qbittorrent for @archlinux) and I assure you if we're modifying detection logic for any sort of dependency we are going to do it in the exported interface of the dependency itself. With a strong preference for pkg-config, but if we absolutely must use `*.cmake` files, we're going to ship it with the dependency itself, NOT with qbittorrent's sources. We also have a strong expectation that if a package provides both `*.cmake` files and pkg-config files, either one must be sufficient to link to it.

So please don't tell me that your advice to refactor these lines into their own file is motivated by your desire to do me a personal favor. It's completely pointless because we don't care one way or the other, and never will.
Please don't even tell me that's *one* of your motivations.

Reply:
@zeule 

> , everyone, I believe, has a right to produce bullshit and name their own or somebody else one to their liking.

You have the right to say whatever you want. But what you said was uncalled for in the context/moment that you said it. Thus far we had a nice back-and-forth going, where I have been listening to every single criticism. Then you took a piece of code that you had never indicated you had a problem with previously, and called it "bullshit you wrote", without even asking what the motivation for it could have been or suggesting changes beforehand. It came across as a hostile action, which will can only be met with hostility.

> > I do understand encapsulation and want to move towards it.
> 
> Then do it, please.

Ok. How many times do I have to tell you that will be done **after this PR**?

> > I have already told you that truly encapsulating the logic in qBittorrent's CMake scripts will happen in a later PR. There are enough changes as it is for now,
> 
> ... ruining encapsulation :)

Do you think it will be harder to move towards more encapsulated logic _after_ these changes? How can that be? Previously there was even global `CXX_FLAGS` manipulation going on, how were the previous scripts easier to work with for this future purpose?

> > More surprises.
>
> Well, I use Windows only to launch Steam. Have no idea what are the best practices there.

Same here. There is only one machine where I have Windows installed on bare metal, and even then it's a dual-boot configuration with Ubuntu. I use that Windows install just for certain games that don't run well (or at all) with Proton/DXVK. But recently I also setup a Windows VM in Ubuntu, as a clean environment for building and testing qBittorrent on Windows. This is because I recognize that the majority of qBittorrent's user base is on Windows, and so whether we like it or not, improving the building process, release automation, etc for that OS is essential for the project. So, in addition to testing these changes on Linux, I'm doing my best to test them thoroughly on Windows.

But this is, in principle, not a potential problem exclusive to Windows anyway. Pulling definitions out of thin air and not telling the user if that happens, as if they had been found normally, is for sure a bad practice no matter the system. Especially when the library in question is so picky about them, as is the case with `libtorrent`.

> > So how about actually explaining calmly what is wrong with that part?
> 
> It fails if PkgConfig finds no libtorrent package,

That's exactly what is supposed to happen, but the "fake package" code is not even reached, the build fails right at `pkg_check_modules`:

```
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
-- Checking for module 'libtorrent-rasterbar>=1.2.0'
--   No package 'libtorrent-rasterbar' found
CMake Error at /usr/share/cmake-3.17/Modules/FindPkgConfig.cmake:467 (message):
  A required package was not found
Call Stack (most recent call first):
  /usr/share/cmake-3.17/Modules/FindPkgConfig.cmake:647 (_pkg_check_modules_internal)
  src/CMakeLists.txt:6 (pkg_check_modules)
```

> it fails if libtorrent version is too new and unsupported.

Good, qBittorrent does not support libtorrent >1.2.x at the moment (in fact, in practice it _only_ supports 1.2.x at the moment). In the future this can be changed of course, as the supported versions change.

> What's wrong with it is that it creates problems because you don't want to follow the standard way for locating packages.

I'm relying on finding a suitable `LibtorrentRasterbarConfig.cmake`, which is as standard as it gets, no? The pkgconfig fallback is a convenience for linux distros that don't build libtorent with CMake (which means no `LibtorrentRasterbarConfig.cmake` is produced), because they will almost for sure have pkgconfig at least. And even still, the user may place a custom `FindLibtorrentRasterbar.cmake` in the project's module directory and that will be used instead.

In fact,the pkgconfig fallback was added at the request of @eli-schwartz (https://github.com/qbittorrent/qBittorrent/pull/12746#issuecomment-623749679). It seemed reasonable enough, and to be honest, I think if an actual package maintainer from a "big"/"mainstream" distribution (who actually packages qBIttorrent, no less) agrees that this is the right way to do it (and further justifies why this is, https://github.com/qbittorrent/qBittorrent/pull/12746#discussion_r426808317, because one should not blindly follow/trust an "authoritative" source), you'll need to provide really convincing counter-arguments to go against that, or find another package maintainer who can do so.

Reply:
> @zeule You're doing a truly terrible job of communicating your intentions.

Sorry for that.
> I told you I was interpreting your statements as mutually contradictory, asked you to clarify your meaning as to what "the point" is, and you simply replied with some "True. True. True." gimmick which I can only assume is supposed to be terribly witty.

I see no contradiction and I thought from my reply it was clear.

> And apparently you also cannot stand to explain yourself to @FranciscoPombal either, so once again instead of taking the opportunity of having a discussion to, well, "discuss" things, you just smear insults across the screen.

For a discussion to take place the subject has to be of interest for both sides

> I'm left wondering why you bother to be here.

The Internet is full of vagabonds, why can't I be one of them?

> This is, to my understanding, nonsense, because anything used to replace said logic would trigger the find_package() invocation.

Let me recollect what we were talking about. There are two points that are related to the topic: 1) append local module path to the directory list and 2) put the whole chunk of the "find installed libtorrent and create imported target" into a special file, called "FindLibtorrentRasterbar.cmake".

The first one is reasoned by the replacement logic, because otherwise local files will be used, and I can still find a dozen of packages in the Gentoo tree that require actual removing of their in-tree Find modules in order to compile.

The second point is rational due to the general principles of managing things in software, because of the way the find_package is implemented in CMake, and because that way is the simplest for distribution. A Find module in CMake, despite the fact that it is usually implemented as a plain code file, is, in fact, a function with the convention about its interface. Its input parameters are well defined, and its output for CMake is also standard (albeit extensible). The body of this function, except trivial cases, is rather complex because it has to handle a lot of nuances. Most of these nuances arise from various distribution methods, naturally, and are a dead-weight in any given case. Therefore this code has to be encapsulated under the well defined interface of `find_package()`.  Why is this helpful for distributions too? Because the place where you need to look for dependencies looks clean and simple.

> On the other hand, it is nonsense anyway, because I AM A LINUX DISTRIBUTION (I package qbittorrent, among other things, for @archlinux) and I assure you if we're modifying detection logic for any sort of dependency we are going to do it in the exported interface of the dependency itself.

Yes, I understand that and just a few days ago submitted a PR to libtorrent to make it install its Find module. 

> So please don't tell me that your advice to refactor these lines into their own file is motivated by your desire to do me a personal favor.

As you wish, but I use the Linux ecosytem too, and care for it.

> It's completely pointless because we don't care one way or the other, and never will.

Yeah, it shows. The ABS is the least functional package build system I ever used. But again, after Portage and Exheres nothing impresses.

Reply:
@FranciscoPombal 
>  But what you said was uncalled for in the context/moment that you said it. 

If you take that personal, I beg your pardon.

> Then you took a piece of code that you had never indicated you had a problem with previously, and called it "bullshit you wrote"…

This is not true, you were told by me and @glassez to wrap that part into a file (where you, so to say,  took it from).

> …which will can only be met with hostility.

I see no problem in that. You are not going to kill somebody because of that, right?

> Ok. How many times do I have to tell you that will be done **after this PR**?

PR are supposed to bring improvements, right? But here it is a step back in this regard for no reason.

> Do you think it will be harder to move towards more encapsulated logic _after_ these changes? How can that be? Previously there was even global `CXX_FLAGS` manipulation going on, how were the previous scripts easier to work with for this future purpose?

I do not defend that code. I just see no reason to destroy other things when you don't have to.

> > Well, I use Windows only to launch Steam. Have no idea what are the best practices there.
> 
> Same here. There is only one machine where I have Windows installed on bare metal, and even then it's a dual-boot configuration with Ubuntu. I use that Windows install just for certain games that don't run well (or at all) with Proton/DXVK. But recently I also setup a Windows VM in Ubuntu, as a clean environment for building and testing qBittorrent on Windows. This is because I recognize that the majority of qBittorrent's user base is on Windows, and so whether we like it or not, improving the building process, release automation, etc for that OS is essential for the project. So, in addition to testing these changes on Linux, I'm doing my best to test them thoroughly on Windows.

This is great, of course. I'm absolutely ignorant with regard to development trends in the Windows word and no VM can help me...

> But this is, in principle, not a potential problem exclusive to Windows anyway. Pulling definitions out of thin air and not telling the user if that happens, as if they had been found normally, is for sure a bad practice no matter the system. Especially when the library in question is so picky about them, as is the case with `libtorrent`.

Exactly because of that there was this line in the Find module:
```cmake
message(STATUS "libtorrent definitions: ${LibtorrentRasterbar_DEFINITIONS}")
```
> That's exactly what is supposed to happen if no suitable pkgconfig file is found, but the "fake package" code is not even reached in that case anyhow, the build fails right at `pkg_check_modules`:
> 
> ```
> -- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
> -- Checking for module 'libtorrent-rasterbar>=1.2.0'
> --   No package 'libtorrent-rasterbar' found
> CMake Error at /usr/share/cmake-3.17/Modules/FindPkgConfig.cmake:467 (message):
>   A required package was not found
> Call Stack (most recent call first):
>   /usr/share/cmake-3.17/Modules/FindPkgConfig.cmake:647 (_pkg_check_modules_internal)
>   src/CMakeLists.txt:6 (pkg_check_modules)
> ```

Exactly. What to do in that case? Windows users have no pkg-config and might have no LibtorrentRaterbarConfig.cmake. MacOS too. 

> I'm relying on finding a suitable `LibtorrentRasterbarConfig.cmake`, which is as standard as it gets, no?

Yes, but it's implementation is what is non-standard in case of libtorrent.

> And even still, the user may place a custom `FindLibtorrentRasterbar.cmake` in the project's module directory and that will be used instead.

I don't think it's a way to go. Why do I need to write such a file from scratch when there are only three bits are required: library location, includes location, the set of configuration defines? You can easily provide a template for this case.

> In fact,the pkgconfig fallback was added 

It was there, you deleted it first to add later.

> at the request of @eli-schwartz ([#12746 (comment)](https://github.com/qbittorrent/qBittorrent/pull/12746#issuecomment-623749679)). It seemed reasonable enough, and to be honest, I think if an actual package maintainer from a "big"/"mainstream" distribution (who actually packages qBIttorrent, no less) agrees that this is the right way to do it (and further justifies why this is, [#12746 (comment)](https://github.com/qbittorrent/qBittorrent/pull/12746#discussion_r426808317), because one should not blindly follow/trust an "authoritative" source), you'll need to provide really convincing counter-arguments to go against that, or find another package maintainer who can do so.

May I skip this, please? Thankfully, libtorrent will install a proper find module with pkg-config check.

Reply:
@zeule 

> @FranciscoPombal
> 
> > But what you said was uncalled for in the context/moment that you said it.
> 
> If you take that personal, I beg your pardon.
> 
> > Then you took a piece of code that you had never indicated you had a problem with previously, and called it "bullshit you wrote"…
> 
> This is not true, you were told by me and @glassez to wrap that part into a file (where you, so to say, took it from).
> 
> > …which will can only be met with hostility.
> 
> I see no problem in that. You are not going to kill somebody because of that, right?

So funny xD. Sad to see a classic "I'm sorry, BUT...". We both know when you told me to "wrap it in a find module", `it` referred to the whole chunk of code. The "fake package" part is just a cosmetic part of it (ultimately not really needed), that makes the information that pkgconfig actually found the package more visible to the user. Otherwise, the pkgconfig output disappears in the scrollback. You then claimed it did not work in 2 cases, and I have proved that those are not a concern (and not a consequence of the "fake package" code itself anyway).

> > Ok. How many times do I have to tell you that will be done **after this PR**?
> 
> PR are supposed to bring improvements, right? But here it is a step back in this regard for no reason.

> > Do you think it will be harder to move towards more encapsulated logic _after_ these changes? How can that be? Previously there was even global `CXX_FLAGS` manipulation going on, how were the previous scripts easier to work with for this future purpose?
> 
> I do not defend that code. I just see no reason to destroy other things when you don't have to.

There are no steps backward in terms of encapsulation in this PR. If anything, there are steps forward (for example, previously, the function for "adding features" also added compile definitions at the source directory level as a result, which is a clear violation of encapsulation in my book; this is no longer the case). But it wouldn't even matter if there were no "steps forward". This PR could very well be just a side-grade in terms of encapsulation. The whole point of it is to simplify the whole foundation so that further encapsulation work is made easier for later.

> This is great, of course. I'm absolutely ignorant with regard to development trends in the Windows word and no VM can help me...

In case you ever need this, here goes some advice: Microsoft provides ready-to-go VMs with all the relevant development software installed (VS, WSL, ...). You may need to use the VS GUI to install some toolchain packages/customize the toolchain packages, but after that, you don't need to use VS or any GUI at all if you don't need it for developing/building. Granted, using the VS dev console (which is just a shortcut that opens a cmd window with all of the relevant PATHs pre-set) is not the best experience, but it does the job. Hopefully in the future they will flesh out Powershell Core (there is already some support for plain Powershell as the dev console) + Windows Terminal support. For (much) better build performance, I recommend fully disabling the anti-virus (there are tutorials online for this; requires fiddling with the group policy editor, turning it off in the anti-virus GUI is of course not enough and it just turns itself back on next boot anyway because it knows better than you) and the search indexer (searching in the explorer GUI sucks anyway, and you can use classic Linux tools via WSL or even powershell for all your searching needs).

> Exactly because of that there was this line in the Find module:
> 
> ```cmake
> message(STATUS "libtorrent definitions: ${LibtorrentRasterbar_DEFINITIONS}")
> ```

Better than nothing for sure, but in my option, I would go more in the direction of:

```cmake
message(WARNING "Failed to obtain library interface metadata for libtorrent, falling back to a default set of libtorrent definitions: ${LibtorrentRasterbar_DEFINITIONS}\nThese definitions may not be (entirely) appropriate for your case! Override by doing xyz ... etc ...")
```

or even fail the build and force the user to pass an option allowing it to continue (as well as overrides to the default set, optionally).

> 
> > That's exactly what is supposed to happen if no suitable pkgconfig file is found, but the "fake package" code is not even reached in that case anyhow, the build fails right at `pkg_check_modules`:
>
> Exactly. What to do in that case? Windows users have no pkg-config and might have no LibtorrentRaterbarConfig.cmake. MacOS too.

Windows users have no incentive not to build with CMake if there is a choice, especially with stuff like `vcpkg` nowadays. Windows users have to compile everything from source anyway, what is the difference if they see the CMake way in the wiki page vs. the b2 way? They are just going to follow the steps. Both Windows and macOS decided to be special snowflakes in a lot of ways when it comes to building software especially, it is perfectly reasonable to just force users of those systems to "just do it this way, it's what works".

> > I'm relying on finding a suitable `LibtorrentRasterbarConfig.cmake`, which is as standard as it gets, no?
> 
> Yes, but it's implementation is what is non-standard in case of libtorrent.

Can you please elaborate on this? I have found it to work quite well, providing the expected information.

> > And even still, the user may place a custom `FindLibtorrentRasterbar.cmake` in the project's module directory and that will be used instead.
> 
> I don't think it's a way to go. Why do I need to write such a file from scratch when there are only three bits are required: library location, includes location, the set of configuration defines? You can easily provide a template for this case.

You don't need to write a basic template yourself. You just landed a PR in libtorrent that makes it so whichever way libtorrent is installed, the example find module is installed as well: so that will be used if nothing else works (I don't think you even need to copy it over manually to qBittorrent's module directory).

> > In fact, the pkgconfig fallback was added
> 
> It was there, you deleted it first to add later.

True, but more importantly, I only added the pkgconfig back later, not everything else in the find script as well.

Reply:
> So funny xD. Sad to see a classic "I'm sorry, BUT...". We both know when you told me to "wrap it in a find module", `it` referred to the whole chunk of code.

Well, if you want to play that game...

https://github.com/qbittorrent/qBittorrent/pull/12746#discussion_r426101657
https://github.com/qbittorrent/qBittorrent/pull/12746#discussion_r426282201
https://github.com/qbittorrent/qBittorrent/pull/12746#discussion_r426286364
https://github.com/qbittorrent/qBittorrent/pull/12746#discussion_r426782808

The fake package" part is the result of your decision to put the find- logic into the main script. 

> You then claimed it did not work in 2 cases, and I have proved that those are not a concern.

You just rejected those cases without any rationale.

> In case you ever need this, here goes some advice from what I have learned so far…

Thank you.

> Better than nothing for sure, but in my option, I would go more in the direction of:

Perfect.

> Windows users have no incentive not to build with CMake…

Let me summarize: there was a working build script that was able to handle various environments and build the project on all the three major OSes. You propose to replace that with a new one that is unable to handle MacOS (can't create bundles), limits Windows users to a single installation method, complicates structure of the main build scrip. Does it gove something in return? I can't see anything. Am I missing something?

> True, but more importantly, I only added the pkgconfig back later, not everything else in the find script as well.

And that's the problem.

Reply:
>pkg_check_modules() is an **alternative** to find_package(), they're two completely different ways to define a cmake "imported target".

Although I am not an advanced CMake user, it seems obvious and logical to me that `pkg_check_modules()`  is one of the ways of finding a certain package and getting information about it, whereas `find_package()`  is a generalized interface between the main build script and the implementation of searching for a specific package (which can have arbitrarily complex logic and use different methods, including pkgconfig).
But if CMake really involves using several (so-called) alternative methods (each of which is not able to do all the work on its own) then I'm even more disappointed in CMake.

Reply:
@zeule 

> The fake package" part is the result of your decision to put the find- logic into the main script.

Do you just want me to remove the "fake pakage logic" completely?

> > You then claimed it did not work in 2 cases, and I have proved that those are not a concern.
> 
> You just rejected those cases without any rationale.

https://github.com/qbittorrent/qBittorrent/pull/12746#discussion_r426901309

> Let me summarize: there was a working build script that was able to handle various environments and build the project on all the three major OSes. You propose to replace that with a new one that is unable to handle MacOS (can't create bundles), limits Windows users to a single installation method, complicates structure of the main build scrip. Does it gove something in return? I can't see anything. Am I missing something?

- Structure of the main buildscript is much simpler now overall. In fact, in what concerns "finding libtorrent", it would just be a one-liner if it did not have the pkgconfig fallback.
- Windows users were already limited to a single (worse) installation method in practice. See the current "Windows compilation" tutorial in the wiki. Plus it's not my fault that Windows does not use `pkgbuild`, otherwise it would not be limited to a single installation method with the new script.
- It works just fine in macOS: see this comment, and previous related ones: https://github.com/qbittorrent/qBittorrent/pull/12746#issuecomment-627300664

> > True, but more importantly, I only added the pkgconfig back later, not everything else in the find script as well.
> 
> And that's the problem.

I did this because I found the previous find script problematic, for the reasons repeated above multiple times. Including a fin script with qBittorrent is just more maintenance burden. We will have to keep it updated with the upstream ourselves. Seems worse than just automatically trusting the `LibtorrentRasterbarConfig.cmake` or pkgcofig files from upstream.

---

Do you think we should include our own `FIndOpenSSL.cmake` as well? Because that makes as much sense as what you're asking for. The current code already accommodates all of the common cases:

1. CMake will first try to find `libtorrent` in `CONFIG` mode, because of `set(CMAKE_FIND_PACKAGE_PREFER_CONFIG ON)` in the top level CML. This means it will try to use auto-generated `LibtorrentRasterbarConfig.cmake` first if possible, which is what we want.
2. If that fails, it re-tries in `MODULE` mode, which means it will use whatever `FindLibtorrentRasterbar.cmake` scripts it can find (if available). This could be an entirely custom one, or a default one provided by `libtorrent` itself, _that we don't have to worry about maintaining_.
3. If that fails, it tries the `pkgconfig` method if available, because it is so commonly available in Linux distros.

Now you might ask, "since the default `FindLibtorrentRasterbar.cmake` already happens to search with `pkgconfig`, isn't the 3rd step redundant?". The answer is yes, it is redundant, but only if you happen to use the default `FindLibtorrentRasterbar.cmake`, or a custom one that does the the `pkgconfig` search. Such a script might not be available, however, or the user may not want to write it, so the `pkgconfig` method serves as a convenience fallback on Linux.

We're doing exactly the same procedure with OpenSSL, except that since `FindOpenSSL.cmake` is guaranteed to exist and suit our needs, we just need to `find_package(OpenSSL)`, which really means we only follow the first 2 steps listed above.

Reply:
Becasue FindOpenSSL.cmake is guaranteed to exist, while FindLibtorrentRasterbar.cmake will appear with libtorrent 1.2.7 (?). When the minimal required libtorrent version for qBittorrent  increases to 1.2.7 or newer, you can drop the local copy of FindLibtorrentRasterbar.cmake. So simple.

Reply:
@glassez 

> Although I am not an advanced CMake user, it seems obvious and logical to me that `pkg_check_modules()` is one of the ways of finding a certain package and getting information about it, whereas `find_package()` is a generalized interface between the main build script and the implementation of searching for a specific package (which can have arbitrarily complex logic and use different methods, including pkgconfig).
> But if CMake really involves using several (so-called) alternative methods (each of which is not able to do all the work on its own) then I'm even more disappointed in CMake.

I also think it makes sense to abstract  `pkg_check_modules()` away in a find script, so that it all happens behind the scenes with a single call to `find_package()`, but one must be aware of the trade-offs/consequences:

- It is either guaranteed that the `FindPakcageXYZ.cmake` script will always be available
or
- You, as the client package, take up the burden and technical debt of bringing along and keeping up-to-date a (possibly customized) copy of some reference `FindPakcageXYZ.cmake`. Or if there is no reference, roll your own.

For libtorrent, we have the following case:
- It is not guaranteed that `FindLibtorrentRasterbar.cmake` exists everywhere.
- I don't think we should carry our own copy of `FindLibtorrentRasterbar.cmake` and worry about it.

That's why I think it's reasonable to follow the 3 steps listed in https://github.com/qbittorrent/qBittorrent/pull/12746#discussion_r428082853.

Reply:
@zeule 

> Becasue FindOpenSSL.cmake is guaranteed to exist, 

Yes, that's what I said.

> while FindLibtorrentRasterbar.cmake will appear with libtorrent 1.2.7 (?). When the minimal required libtorrent version for qBittorrent increases to 1.2.7 or newer, you can drop the local copy of FindLibtorrentRasterbar.cmake. So simple.

This code works great with or without libtorrent >=1.2.7. I don't want to wait until the minimum required version is 1.2.7.

Reply:
So, @glassez, what is your verdict?

Reply:
I didn't review it completely... will do it later.
But this place seems bad to me. Maybe, of course, this is how it should be in CMake... (I'm not an advanced CMake user).

Reply:
Well, I finally found the time to test it... and I failed it! More precisely, my test failed from the first steps. I found another drawback of these changes that is "blocking issue" for me.
Your patch broke my development scenario. I need to do parallel development of qBittorrent and libtorrent (I'm making changes to both projects), so I don't have an "installed" libtorrent, but only a configured project in the IDE (QtCreator) so my main project depends on it and build it first. It's very easy to configure there (especially in the case of QMake, nothing else is required).
The initial problem with CMake was that It didn't allow you to configure projects because it couldn't find libtorrent. For the current CMake script, it was easy to get around this inconvenience, I only had to set a couple of CMake variables in the project settings (LibtorrentRasterbar_INCLUDE_DIR and LibtorrentRasterbar_LIBRARY), so it didn't search for the package, but just used these values. With this changes I can't use this solution. I get the following message:
>D:\Users\glassez\Documents\Projects\qbittorrent\qBittorrent\src\CMakeLists.txt:21: error: By not providing "FindLibtorrentRasterbar.cmake" in CMAKE_MODULE_PATH this project has asked CMake to find a package configuration file provided by "LibtorrentRasterbar", but CMake did not find one. Could not find a package configuration file provided by "LibtorrentRasterbar" (requested version 1.2.0) with any of the following names: LibtorrentRasterbarConfig.cmake libtorrentrasterbar-config.cmake Add the installation prefix of "LibtorrentRasterbar" to CMAKE_PREFIX_PATH or set "LibtorrentRasterbar_DIR" to a directory containing one of the above files.  If "LibtorrentRasterbar" provides a separate development package or SDK, be sure it has been installed.

I'm sorry, I can't approve of it until it is fixed or another solution is presented.

Reply:
Well, seems it is possible to make libtorrent a part of qBittorrent project itself just by include it. Am I right? Then we just need to introduce some variable, e.g. LibtorrentRasterbar_PROJECT_DIR, and handle it (when it's set skip package searching and include libtorrent project).
Can you investigate into it? Or do you think about another solution? (I can test it when I have time.)

Reply:
> Unfortunately the above won't work...

have you considered https://cmake.org/cmake/help/latest/module/ExternalProject.html

Reply:
>have you considered

No. I don't know much about CMake. And I don't have enough time to study it in detail.
I can try it, of course, but I'm not sure it will work easily. Judging by the amount of documentation, it is not so easy to configure... in addition, this is the job of someone who broke the handling of the use case I described above ("satellite projects" development).

Reply:
Assuming Libtorrent is built with cmake too, one can supply `LibtorrentRasterbar_DIR` variable when configuring qBt. The variable has to point to the directory containing the LibtorrentRasterbarConfig.cmake file (<build dir>/LibtorrentRasterbar). This works with any CMake `find_package()` call (https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure).

Reply:
>one can supply LibtorrentRasterbar_DIR variable when configuring qBt. The variable has to point to the directory containing the LibtorrentRasterbarConfig.cmake file

It won't work in described use case.
What I need is the ability to tell CMake that some dependent package is not installed at the time of project configuration, but should be compiled from sources before building the project (or at least trick CMake into believing that the package was found, as I did earlier, and allow QtCreator to handle dependency between projects).

Reply:
I can re-assure you it works exactly as you want, but you need to run configure for libtorrent manually for the first time.

Reply:
>you need to run configure for libtorrent manually for the first time.

I know it already but it's exactly what I don't want!
I just want to perform "Build" on qBittorrent project and allow my IDE to resolve its dependencies. This is something that works perfectly with the native Qt build system (QMake). And it should continue to work with CMake on the same principle (configured and forgot).

Reply:
libtorrent does not provide qmake build script. If the build works for in the way you describe, you must have wrote a qmake rule to build libtorrent. If so, why don't you do the same for CMake, the mainstream C++ build system?

Reply:
I don't think you understand what I'm talking about...
Here is a description of my workflow (two ways, common things are omitted):
1. Create "Solution" in QtCreator and add two "Project": CMake-based "libtorrent" and "qBittorrent" (either QMake or Cmake based).
2. Configure "qBittorrent" to use "libtorrent" include dir and built library (using appropriate variables in QMake based project or LibtorrentRasterbar_INCLUDE_DIR/LibtorrentRasterbar_LIBRARY in CMake based project - before this PR it also prevents CMake to search for libtorrent package).
3. Set "libtorrent" project as dependency for "qBittorrent" project (in QtCreator project settings). So from now on QtCreator handle this dependency.

A piece of cake! And I want it to stay that way.

Reply:
I can't see how that differs from the workflow with PackageName_DIR that I proposed to you above. As an illustration here was my setup with KDevelop:
1. Add libtorrent cmake project to a new session or a session with qBt project.
2. Configure the libtorrent project to make it generate export targets and package config files in the LibtorrentRasterbar subdir of its build dir.
3. If the qBt project was already configured and picked the system libtorrent up, either remove Libtorrent variables from the project cmake cache or delete the cache or build dir altogether.
4. (Re-)Configure the qBt project passing `-DLibtorrentRasterbar_DIR=<libtorrent_build_dir/LibtorrentRasterbar/>`
5. Ensure libtorrent is placed earlier than qBt In the session build target list.

Looks pretty match the same for me. Or you can go the formal way and create a superbuild project using the ExternalProject commands, as it was already suggested in the thread.

Reply:
@glassez 

Sorry for not responding earlier, but this thread has gotten quite long and this particular comment chain is starting to get buried, so I only saw this now.

As mentioned, if you've built libtorrent with CMake, it is quite easy to use it when building qBittorrent with CMake, whether you "install" it or not. According to what I understood of your workflow, you  want to build libtorrent and use the files directly from the build dir without "installing it". To accomplish this, you can simply point `LibtorrentRasterbar_DIR` to the libtorrent build dir.

For example, if you've built libtorrent like so:

```
cd libtorrent
cmake -S . -B cmake-build-dir -DCMAKE_CXX_STANDARD=14 -DCMAKE_BUILD_TYPE=RelWithDebInfo -Ddeprecated-functions=FALSE
```

All the build files will go to `libtorrent/cmake-build-dir`, and will be usable in this state (including the CMake config/target scripts), even before you issue the install command.
In particular, `LibtorrentRasterbarConfig.cmake` will be inside `libtorrent/cmake-build-dir/LibtorrentRasterbar`

So, when building qBittorrent, you just need to add: `-DLibtorrentRasterbar_DIR="/absolute/path/to/libtorrent/cmake-build-dir/LibtorrentRasterbar"` to its CMake configure command line.

I imagine setting it up like this in QtCreator should be no problem.

Reply:
>I imagine setting it up like this in QtCreator should be no problem.

It's not a problem. The problem is that I require some additional manual actions since this PR... although it's not so big deal. Seems Qt Creator has its own drawbacks (about integration with CMake), so maybe I need to send a couple of bug reports there first.

Well, even if you don't look at it Qt Creator integration, there is a more general/serious problem here, which is the inability to use it on Windows if libtorrent isn't built with CMake (please correct me if I'm wrong).

Reply:
@glassez 

> It's not a problem. The problem is that I require some additional manual actions since this PR... although it's not so big deal. Seems Qt Creator has its own drawbacks (about integration with CMake), so maybe I need to send a couple of bug reports there first.

Yes, conceptually it's not even a big deal; it's just "when building CMake project X (qBittorrent) with some specified options, build CMake project Y (libtorrent) first with some specified options". Since Qt itself is moving to CMake very soon, I would expect a basic flow like this to be easily configurable, if not today, then in the near future.

> Well, even if you don't look at it Qt Creator integration, there is a more general/serious problem here, which is the inability to use it on Windows if libtorrent isn't built with CMake (please correct me if I'm wrong).

I'm not sure how libtorrent built with `b2` was propagating usage requirements to qBittorrent's qmake build previously. But now, I think the best solution for this is to have libtorrent generate/bundle CMake config files even when built with `b2`, so that it can easily be used in any CMake project. libtorrent is very sensitive to the flags used in its compilation and dependent projects, so automatic usage requirement resolution and propagation is a must. I honestly don't expect anyone building qBittorrent (or any C++ project based on libtorrent) to want to add libtorrent's libdir, include dir, compile flags, defines, etc manually.

An existing example of this, actually relevant to us: right now, Qt is built with qmake, but it still generates/bundles proper CMake config files. That's why you can build Qt yourself (like in the wiki, with qmake) and have it still "just work" in CMake projects:

> - Compiled qBittorrent itself using the manually-built [following the wiki procedure] dependencies:
    ```
    cd qBittorrent
    cmake -S . -B build -DVERBOSE_CONFIGURE=ON -DCMAKE_PREFIX_PATH=G:\QBITTORRENT\install_msvc64\base -DBoost_ROOT=G:\QBITTORRENT\boost_1_72_0\stage -DBoost_USE_STATIC_RUNTIME=ON -DQt5_DIR=G:\QBITTORRENT\install_msvc64\qt5\lib\cmake\Qt5
    cmake --build .\build\ --config RelWithDebInfo
    ```
>
>and... it worked! (...)

Note the `-DQt5_DIR=G:\QBITTORRENT\install_msvc64\qt5\lib\cmake\Qt5`.

But yes, at present, you're right. Until libtorrent's `b2` build generates CMake config files, it is not feasible to build qBittorrent on Windows with CMake if libtorrent wasn't also built with CMake.

Reply:
> But yes, at present, you're right. Until libtorrent's `b2` build generates CMake config files, it is not feasible to build qBittorrent on Windows with CMake if libtorrent wasn't also built with CMake.

Because you broke that in this very PR. LOL

Reply:
>I'm not sure how libtorrent built with b2 was propagating usage requirements to qBittorrent's qmake build previously. 
...
libtorrent is very sensitive to the flags used in its compilation and dependent projects, so automatic usage requirement resolution and propagation is a must. I honestly don't expect anyone building qBittorrent (or any C++ project based on libtorrent) to want to add libtorrent's libdir, include dir, compile flags, defines, etc manually.

You seem to be exaggerating... there is no such big problem here (at least for developers). I've always done it manually. What's the problem with using the same settings to build these two projects?
Actually, CMake gave me more trouble when I had to make it work correctly (e.g. when it failed to find Boost in the folder where it was without any helpful error messages).

In any case, I see no serious reason to break supporting libtorrent that is built with something other than CMake (at least as part of current changes).

Reply:
@zeule

Let's see what I "broke" (quote from the FindLibtorrentRasterbar.cmake file deleted by this PR):

>```cmake
># Without pkg-config, we can't possibly figure out the correct build flags.
># libtorrent is very picky about those. Let's take a set of defaults and
># hope that they apply. If not, you the user are on your own.
>```

Which is of course an excellent idea - set some important parameters to some values and hope for the best silently, without even informing/warning the user.

Moving to an objectively better solution is not "breaking", it's "improving". It just so happens that an improvement is also necessary in libtorrent's side (`b2` builds need to generate CMake config files).

I'd like to hear how asking for a library to generate config files for the standard C++ build system  when using another build system is not preferable to the previous solution. Perhaps we should also open a bug report on Qt's bug tracker saying "hey guys, you know that thing you were doing with the qmake build generating proper, usable cmake config files? scratch that, just publish a findQt5.cmake file, tell everyone that they are expected to copy paste it in their projects (and to keep an eye out for changes :)). Oh, and make sure to silently set important variables to arbitrary values to what you think are sane when whatever automatic detection methods you have in place fail, without warning the user.".

Reply:
There are two calls in the FindLibtorrentRasterbar.cmake file that make it possible to use it if you are on Windows and libtorrent is built with boost build (or in any other situation when the cmake way could not work): `find_path()` and `find_library()` with hints. 

If you do not understand the purpose of the code are you throwing out, maybe it's time to pause for a while and think a little bit?

As to your comment about random libtorrent flags, I can only say it is unrelated to the topic. The closest related think is the`LibtorrentRasterbar_CUSTOM_DEFINITIONS` variable in the same file.

Reply:
At least current (master) approach allows you to use any libtorrent distribution and I don't think this is a flaw that needs to be fixed.
Please extract this change into separate PR if you want to continue this discussion.

Reply:
@zeule 

> There are two calls in the FindLibtorrentRasterbar.cmake file that make it possible to use it if you are on Windows and libtorrent is built with boost build (or in any other situation when the cmake way could not work): `find_path()` and `find_library()` with hints.
> 
> If you do not understand the purpose of the code are you throwing out, maybe it's time to pause for a while and think a little bit?

I never claimed there was no way of finding libtorrent before, I claimed that my changes move towards a better way of doing it, that will also require changes on the libtorrent side. If these changes land before the required ones in libtorrent, then yes, there will be a period of time where building libtorrent without CMake on Windows won't work for building qBittorrent. But who cares? Window users can just build it with CMake in the meantime! CMake comes bundled with VS so most people interested in building on Windows already have it, and even if it didn't, Windows users are no strangers to the simple task of downloading an exe from the official website (and optionally adding it to PATH). Not to mention that in the future, they'll have to have CMake anyway already to build qBittorrent. Using CMake to build both projects is the most practical way already.

You have already submitted a PR in libtorrent that touched some `b2` files, so you probably have some experience with that. Additionally, you also recognize the importance of CMake in the C++ ecosystem. So, instead of arguing with me here, why not spend that time implementing the better solution over there (making libtorrent `b2` builds generating CMake config files)? Personally, I have no interest in doing that myself, I'm happy to just tell Windows users "just build it with CMake", and I doubt the vast majority would take issue with that or let it bother them.

> As to your comment about random libtorrent flags, I can only say it is unrelated to the topic. The closest related think is the`LibtorrentRasterbar_CUSTOM_DEFINITIONS` variable in the same file.

It is relevant. In the old code:

- On Windows, `PC_LIBTORRENT_RASTERBAR_FOUND` will be false (because no `pkgconfig`), leading to the `if(LibtorrentRasterbar_CUSTOM_DEFINITIONS) ... else()` conditional.
    - If the user manually specifies `LibtorrentRasterbar_CUSTOM_DEFINITIONS`, that will be used. But this is not only tedious and error-prone, it is unlikely that the user will even know about `LibtorrentRasterbar_CUSTOM_DEFINITIONS` in the first place, since it's "documentation" is a comment in the script.
    - If the `else()` branch is taken, we fall into the part I mentioned above: `# Without pkg-config, we can't possibly figure out the correct build flags. (...)`, where flags that may not work are set without warning.
- After that, nothing noteworthy happens to `LibtorrentRasterbar_DEFINITIONS`. It is used to determine if OpenSSL is needed or not. Additionally, a few more things may get appended if `LibtorrentRasterbar_USE_STATIC_LIBS` (another important variable documented by comment) is not set.
- It is then added to the target properties. May not work, hope for the best.

Reply:
@glassez 

> You seem to be exaggerating...

Exaggerating about what? `libtorrent is very sensitive to the flags (...)` is no exaggeration.

> there is no such big problem here (at least for developers). I've always done it manually. What's the problem with using the same settings to build these two projects?

You think having to set libtorrent's libdir, include dir, compile flags, defines, etc manually is a good way of doing things? Same with the official qmake build procedure in the wiki (https://github.com/qbittorrent/qBittorrent/wiki/Compiling:-MSVC-2019-(static-linkage)#Compiling_qBittorrent), having to manually edit the build script just to build is insane. Maybe it does not bother you because you've been doing it for a long time, but don't let that cloud your judgement.

> Actually, CMake gave me more trouble when I had to make it work correctly (e.g. when it failed to find Boost in the folder where it was without any helpful error messages).

It's difficult to be simpler than (it can be even simpler with something like `vcpkg` providing a toolchain file):

```
cd libtorrent
cmake -S . -B cmake-build-dir -DCMAKE_CXX_STANDARD=14 -Ddeveloper-options=ON -Ddeprecated-functions=OFF -Dstatic_runtime=ON -DBUILD_SHARED_LIBS=OFF -DCMAKE_PREFIX_PATH=G:\QBITTORRENT\install_msvc64\base -DBoost_ROOT=G:\QBITTORRENT\boost_1_72_0\stage -DCMAKE_INSTALL_PREFIX=G:\QBITTORRENT\install_msvc64\base
cmake --build .\cmake-build-dir\ --config RelWithDebInfo && cmake --install .\cmake-build-dir\ --config RelWithDebInfo

cd qBittorrent
cmake -S . -B build -DVERBOSE_CONFIGURE=ON -DCMAKE_PREFIX_PATH=G:\QBITTORRENT\install_msvc64\base -DBoost_ROOT=G:\QBITTORRENT\boost_1_72_0\stage -DBoost_USE_STATIC_RUNTIME=ON -DQt5_DIR=G:\QBITTORRENT\install_msvc64\qt5\lib\cmake\Qt5
cmake --build .\build\ --config RelWithDebInfo
```

It works for everyone with the same files (no need to edit them), each person just has to change paths like `-DBoost_ROOT=...` to what they're using at the command line. Scriptable, automation-friendly and doesn't require a GUI. If you're missing any important `-D_.*_{ROOT,DIR}` option, you'll get a helpful message like this for each of them (example for boost):

```
error (...)
Please set BOOST_ROOT to the root
directory containing Boost or BOOST_INCLUDEDIR to the directory containing
Boost's headers.
(...)
```

No idea why it failed without a helpful message in your case. Perhaps something wrong with your boost installation?

> In any case, I see no serious reason to break supporting libtorrent that is built with something other than CMake (at least as part of current changes).

It's not a serious problem to break it temporarily, until libtorrent's `b2` build is capable of generating the required files. The workaround, which is "just build it with CMake" is more than sufficient and acceptable IMO.

Reply:
@FranciscoPombal
Are you still going to slow down the entire PR because of this unresolved issue, or are you going to separate it?

There is nothing too problematic for me personally, just some inconveniences...
But can this create problems for users of some more "exotic" systems? I mean, how likely is it that some distribution has installed libraries without both CMake and PkgConfig configuration files?
Do we need to hear more opinions about this particular change?

P.S. I am ready to approve it as soon as my last questions are satisfied (as well as when the pending changes are completed and the commits are squashed).
@Chocobo1?

Reply:
@glassez 

> But can this create problems for users of some more "exotic" systems? I mean, how likely is it that some distribution has installed libraries without both CMake and PkgConfig configuration files?

As far as I know, both of libtorrent's buildsystems produce at least the pkgconfig configuration files. You could argue "what happens in distros that don't have pkgconfig?" - but is this really an issue? I'd say it's very unlikely that there exists a relevant modern distro that doesn't have pkgconfig (although if anyone has hard evidence to the contrary, please enlighten me).

Anyway, what I'm advocating for (making libtorrent's b2 build method also produce CMake configuration files) makes the "not having pkgconfig" problem a non-issue.

Of course, this whole discussion has an even simpler solution: just don't have 2 different buildsystems. That way we wouldn't even have to worry about having the libtorrent b2 build producing CMake config files. Why does libtorrent have to have 2 buildsystems? I suppose it still has both because it started out with b2 and never really dropped it completely, even up to now. If it could only be built with CMake, we would not be having this discussion, everyone would build it happily just fine and no one would complain. But since it happens to be able to be built a different way (which currently does not generate the required config files for the other buildsystem), everyone is sort of forced to support both, because any distro may to decide to use either one.(1)

Supporting 2 different buildsystems is just a waste of time. It's ok to have 2 during some kind of "transition phase" when moving from one to the other, but to keep supporting both indefinitely? Just wasteful.(2) Ultimately, a decision has to be made in favor of one for the project, and then everyone should "get with the program".

(1) We can always just ignore this, pretend libtorrent built with b2 doesn't exist, and say "your libtorrent distro package was built with b2 and it did not produce CMake config files? How terrible. Grab the libtorrent source and compile from source yourself".

(2) There is an an exception to this: when one of the platforms the application must support absolutely requires using a different specific buildsystem than the rest. I don't think that's the case for libtorrent, qBittorrent, or similar projects. Modern CMake alone is more than adequate for these projects.

> P.S. I am ready to approve it as soon as my last questions are satisfied

Besides this one, can you please remind me of the other unresolved concerns you have?

> Are you still going to slow down the entire PR because of this unresolved issue, or are you going to separate it?

We don't have to slow down for this. We could just do it this way. If there are too many complaints, we can backtrack and re-add the package find file. But I doubt there will be many, if any complaints. The complaint so far is hypothetical. Everyone else is either indifferent to this find package change (which is a good thing, it means it "just works"), or actively supporting it (an actual Arch Linux package maintainer, no less). I would consider making package maintainers' lives easier a nice bonus.

Reply:
>Besides this one, can you please remind me of the other unresolved concerns you have?

No, I meant the questions from my previous comment only.

Reply:
@glassez 

This one? https://github.com/qbittorrent/qBittorrent/pull/12746#discussion_r442207877

> At least current (master) approach allows you to use any libtorrent distribution and I don't think this is a flaw that needs to be fixed.
Please extract this change into separate PR if you want to continue this discussion.

You can still use any libtorrent distribution, as long as you have pkgconfig or CMake.

Reply:
@FranciscoPombal
Finally, I have completed my research on this issue.
Fortunately, your "totalitarian" tendencies will not be satisfied, and it still allows you to use any libtorrent source via `FindLibtorrentRasterbar.cmake` script installed in the system (or provided to CMake in any possible way). It looks for CMake configuration files first, and if it fails at this, it still tries to find `FindLibtorrentRasterbar.cmake` file and use it. IMO, it works now even better than it did before, so I am satisfied and this issue is closed for me.

Reply:
I'm confused... wasn't the *entire* point of this change, that it first looks for  `FindLibtorrentRasterbar.cmake`, wherever or however you produce one (e.g. by a system libtorrent-rasterbar install, or if you so wish by providing your own)?

That was why I objected to moving this `find_package()` into a qBittorrent `cmake/Modules/FindLibtorrentRasterbar.cmake`, because then you would have recursion issues.

Reply:
>That was why I objected to moving this `find_package()` into a qBittorrent `cmake/Modules/FindLibtorrentRasterbar.cmake`, because then you would have recursion issues.

Did someone suggest to move `find_package()` there? Perhaps this is some kind of misunderstanding... I suggested to move all "find libtorrent" related stuff there so we have only one call to `find_package(libtorrent)` in main script. Even now I think it is good idea to have it this way (to have only one call to `find_package(libtorrent)` and let someone else to provide some specific way via FindLibtorrentRasterbar.cmake if needed.
>wasn't the entire point of this change, that it first looks for FindLibtorrentRasterbar.cmake

`find_package()` by default tries to find and use corresponding FindPackage.cmake module and if there is no such module it falls back to find and use corresponding CMake configuration files. But it is changed to opposite direction in this PR.

Reply:
> Did someone suggest to move `find_package()` there? Perhaps this is some kind of misunderstanding... 

It is quite literally the review comment by zeule which started this discussion, so yes, someone did indeed suggest precisely that.

> I suggested to move all "find libtorrent" related stuff there

Including the find_package(), which I objected to because of namespacing issues.

> so we have only one call to `find_package(libtorrent)` in main script. Even now I think it is good idea to have it this way (to have only one call to `find_package(libtorrent)` and let someone else to provide some specific way via FindLibtorrentRasterbar.cmake if needed.

There is no difference to people interested in providing some FindLibtorrentRasterbar.cmake if needed. Moving this code to find_package(libtorrent) as opposed to leaving it right here merely runs the same exact code via an included macro of cmake function calls. There is no technical advantage to having it in its own file, there is only a readability advantage.

But naming it "Libtorrent" instead of "LibtorrentRasterbar" may avoid the namespacing/recursion issue I raised, however, it instead opens up the amusing possibility the https://github.com/rakshasa/libtorrent project might add a FindLibtorrent.cmake, and result in possibly finding the wrong thing.

"Naming is hard."

Reply:
If it is that important to move a few lines somewhere else, I suggest `QbtLibtorrentImpl.cmake` and `include(QbtLibtorrentImpl)`.

Reply:
@eli-schwartz, sorry, I can't understand your reasoning. What do you want from it?
IIRC, this new configuration allows you to use any libtorrent package that provide CMake configuration files or other kind of libtorrent distribution if there is FindLibtorrentRasterbar.cmake exists in any known location (or using pkgconfig as 3rd attempt). We can remove pkgconfig way from our script but it can be added by libtorrent/CMake distributors to their FindLibtorrentRasterbar.cmake.
>There is no technical advantage to having it in its own file, there is only a readability advantage.

Are you sure? Having second find_package call for the same package only to make script failed isn't readability issue for me. Looks like workaround.

Reply:
Anyway I suggest to continue this discussion in separate thread unless there is something blocking here.

Reply:
@glassez 

> Fortunately, your "totalitarian" tendencies will not be satisfied,

Not sure what you mean by this... my goal was to have the search procedure working with _at least_ 1) CMake config files and 2) pkg-config. If, in addition, it became possible to use an external `FindLibtorrentRasterbar.cmake` to drive the search procedure, then great (as you confirmed, this is the case), but otherwise, no problem.

What I did not want, in essence, was for us to provide our own `FindLibtorrentRasterbar.cmake` in the repository.

So I guess we all got what we wanted in the end?

> Anyway I suggest to continue this discussion in separate thread unless there is something blocking here.

Agreed, it has become unwieldy to continue the conversation in this thread. Any interested parties can open a new dedicated thread for discussion after this PR is merged, I'll mark this one as resolved now. Thanks for the input.